src/apps/peer/udpserver.c

```diff
@@ -32,86 +32,123 @@
 #include "apputils.h"
 #include "stun_buffer.h"
 
+struct listen_info {
+  ioa_addr addr;
+  int received;
+  int verbose;
+  struct listen_info *next;
+};
+
 /////////////// io handlers ///////////////////
 
 static void udp_server_input_handler(evutil_socket_t fd, short what, void *arg) {
 
   if (!(what & EV_READ))
     return;
 
-  ioa_addr *addr = (ioa_addr *)arg;
+  struct listen_info *listen = (struct listen_info *)arg;
 
   int len = 0;
-  int slen = get_ioa_addr_len(addr);
-  stun_buffer buffer;
-  ioa_addr remote_addr;
+  int slen = get_ioa_addr_len(&listen->addr);
+  stun_buffer *buffer = (stun_buffer *)malloc(sizeof(stun_buffer));
+  ioa_addr remote_addr = {0};
+  if (!buffer)
+    return;
 
   do {
-    len = recvfrom(fd, buffer.buf, sizeof(buffer.buf) - 1, 0, (struct sockaddr *)&remote_addr, (socklen_t *)&slen);
+    len = recvfrom(fd, buffer->buf, sizeof(buffer->buf) - 1, 0, (struct sockaddr *)&remote_addr, (socklen_t *)&slen);
   } while (len < 0 && socket_eintr());
 
-  buffer.len = len;
+  buffer->len = len;
 
   if (len >= 0) {
+    listen->received += len;
+    if (listen->verbose) {
+      uint8_t ra[64] = {0}, la[64] = {0};
+      addr_to_string(&listen->addr, la);
+      addr_to_string(&remote_addr, ra);
+      TURN_LOG_FUNC(TURN_LOG_LEVEL_DEBUG, "Received %d bytes data %s from %s. total received: %d bytes\n", len, la, ra,
+                    listen->received);
+    }
+
     do {
-      len = sendto(fd, buffer.buf, buffer.len, 0, (const struct sockaddr *)&remote_addr, (socklen_t)slen);
+      len = sendto(fd, buffer->buf, buffer->len, 0, (const struct sockaddr *)&remote_addr, (socklen_t)slen);
```

Comment:
Maybe i'm not understanding the design properly, but why are we sending the buffer we just read...? Doesn't the buffer need to be checked for contents of any sort?

Reply:
Unless this particular application is simply intended to spit back the exact buffer it received.
If so, that doesn't seem all that useful, but i could use it in specific test situations.

Reply:
This app is an implementation of "peer" so that tests could be run without relying on external code/clients. Echoing back the received packets is what it does - for tests exactly @jonesmz